audition/distance_from_best.py

```diff
@@ -0,0 +1,300 @@
+                        value as raw_value,
+                        best_val - value below_best
+                    FROM model_tols
+                )
+                select
+                    current_best_vals.*,
+                    first_value(below_best) over (
+                        partition by model_group_id
+                        order by train_end_time asc
+                        rows between 1 following and unbounded following
+                    ) below_best_next_time
+                from current_best_vals
+            '''.format(
+                model_group_ids=str_in_sql(model_group_ids),
+                train_end_times=str_in_sql(train_end_times),
+                models_table=self.models_table,
+                metric=metric['metric'],
+                metric_param=metric['param'],
+                new_table=self.distance_table
+            ))
+
+    def create_and_populate(
+        self,
+        model_group_ids,
+        train_end_times,
+        metrics,
+    ):
+        """Creates and populates the distance table with the
+            given model groups, times, and metrics
+
+        Args:
+            model_group_ids (list) Model group ids to include in the distance table
+            train_end_times (list) Train end times to include in the table
+            metrics (list) Metrics and params to include in the table. Each
+                row should be a dict with keys:
+                        'metric' (e.g. 'precision@')
+                        'param' (e.g. '100_abs')
+                All models should have the results.evaluations table populated
+                for all given model group ids, train end times, and metric/param combos
+        """
+        self._create()
+        self._populate(model_group_ids, train_end_times, metrics)
+
+    def as_dataframe(self, model_group_ids):
+        """Return model-group-id subset of table as dataframe
+
+        Args:
+            model_group_ids (list) the desired model group ids
+
+        Returns: (pandas.DataFrame) The data from the table corresponding
+            to those model group ids
+        """
+        return pd.read_sql(
+            'select * from {} where model_group_id in ({})'.format(
+                self.distance_table,
+                str_in_sql(model_group_ids)
+            ),
+            self.db_engine
+        )
+
+    def dataframe_as_of(self, model_group_ids, train_end_time):
+        """Return model group id/train end time subset of table as dataframe
+
+        Args:
+            model_group_ids (list) the desired model group ids
+            train_end_date (string) the desired train end time
+
+        Returns: (pandas.DataFrame) The data from the table corresponding
+            to those model group ids and train end time
+        """
+        base_df = self.as_dataframe(model_group_ids)
+        return base_df[base_df['train_end_time'] == train_end_time]
+
+
+class BestDistanceHistogrammer(object):
```

Comment:
Might consider renaming to `BestDistancePlotter` or something -- it's not really a histogram (more like a cumulative distribution function)